osu.Game/Database/StandardisedScoreMigrationTools.cs

```diff
@@ -0,0 +1,178 @@
+using osu.Game.Rulesets.Judgements;
+using osu.Game.Rulesets.Objects;
+using osu.Game.Rulesets.Scoring;
+using osu.Game.Scoring;
+
+namespace osu.Game.Database
+{
+    public static class StandardisedScoreMigrationTools
+    {
+        public static long GetNewStandardised(ScoreInfo score)
+        {
+            // Avoid retrieving from realm inside loops.
+            int maxCombo = score.MaxCombo;
+
+            var ruleset = score.Ruleset.CreateInstance();
+            var processor = ruleset.CreateScoreProcessor();
+
+            processor.TrackHitEvents = false;
+
+            var beatmap = new Beatmap();
+
+            HitResult maxRulesetJudgement = ruleset.GetHitResults().First().result;
+
+            var maximumJudgements = score.MaximumStatistics
+                                         .Where(kvp => kvp.Key.AffectsCombo())
+                                         .OrderByDescending(kvp => Judgement.ToNumericResult(kvp.Key))
+                                         .SelectMany(kvp => Enumerable.Repeat(new FakeJudgement(kvp.Key), kvp.Value))
+                                         .ToList();
+
+            // This is a list of all results, ordered from best to worst.
+            // We are constructing a "best possible" score from the statistics provided because it's the best we can do.
+            List<HitResult> sortedHits = score.Statistics
+                                              .Where(kvp => kvp.Key.AffectsCombo())
+                                              .OrderByDescending(kvp => Judgement.ToNumericResult(kvp.Key))
+                                              .SelectMany(kvp => Enumerable.Repeat(kvp.Key, kvp.Value))
+                                              .ToList();
+
+            if (maximumJudgements.Count != sortedHits.Count)
+            {
+                // Older scores may not have maximum judgements populated correctly.
+                // In this case we need to fill them.
+                maximumJudgements = sortedHits
+                                    .Select(r => new FakeJudgement(getMaxJudgementFor(r, maxRulesetJudgement)))
+                                    .ToList();
+            }
+
+            foreach (var judgement in maximumJudgements)
+                beatmap.HitObjects.Add(new FakeHit(judgement));
+
+            processor.ApplyBeatmap(beatmap);
+
+            Queue<HitResult> misses = new Queue<HitResult>(score.Statistics
+                                                                .Where(kvp => kvp.Key == HitResult.Miss || kvp.Key == HitResult.LargeTickMiss)
+                                                                .SelectMany(kvp => Enumerable.Repeat(kvp.Key, kvp.Value)));
+
+            int maxJudgementIndex = 0;
+
+            foreach (var result in sortedHits)
+            {
+                // misses are handled from the queue.
+                if (result == HitResult.Miss || result == HitResult.LargeTickMiss)
+                    continue;
```

Comment:
I'm... not getting how this works at all. How can misses be handled from the queue if this is a `continue`?

Reply:
There's a separate queue with the same items in it. So we're ignoring them in the main loop and inserting them along the way whenever a miss is required.

The reason this is here rather than in the LINQ is because `            if (maximumJudgements.Count != sortedHits.Count)` needs to compare the count, including misses, and I didn't want to do two separate linq queries.